### PR TITLE
Move the canonical class to parsing and core modules

### PR DESCRIPTION
**Pull request by agoscinski**
There is a bug the `Task.plugin_classes` are not yet initialized. Maybe it is because of the constructor approach I take. Need to investigate

### CODE REVIEW

File: tests/unit_tests/parsing/test_yaml_data_models.py [hunk 1]
```diff
@@ -2,20 +2,21 @@
 import textwrap
 
 from sirocco.parsing import _yaml_data_models as models
+from sirocco.core import Workflow
 
 
-def test_workflow_canonicalization():
+def test_workflow_():
```
Comment (DropD) [inline tests/unit_tests/parsing/test_yaml_data_models.py:8]: A test for `Workflow.from_config_workflow` already exists. If this now tests something in addition to the existing one, move it to `test_workflow` (and deduplicate). Else, remove the module for now.

File: src/sirocco/parsing/_yaml_data_models.py [hunk 2]
```diff
@@ -541,11 +548,12 @@ class ConfigWorkflow(BaseModel):
 
     """
 
+    rootdir: None | Path = None
```
Comment (DropD) [inline src/sirocco/parsing/_yaml_data_models.py:551]: We took `rootdir` out of here, because it can not be supplied inside the yaml file, based on Matthieu's comment on #89 . What is the reasoning for reintroducing it here?

File: src/sirocco/parsing/_yaml_data_models.py [hunk 3]
```diff
@@ -570,67 +578,21 @@ def check_parameters(self) -> ConfigWorkflow:
...
-    data: ConfigData
-    parameters: dict[str, list[Any]]
-
-    @property
-    def data_dict(self) -> dict[str, ConfigAvailableData | ConfigGeneratedData]:
-        return {data.name: data for data in itertools.chain(self.data.available, self.data.generated)}
-
-    @property
-    def task_dict(self) -> dict[str, ConfigTask]:
-        return {task.name: task for task in self.tasks}
```
Comment (DropD) [inline src/sirocco/parsing/_yaml_data_models.py:598]: Do you not find the property approach more elegant than duplicating this information manually? If you are worried about performance, it can always be turned into a cached property, once we have evidence that it is actually problematic.

File: src/sirocco/parsing/_yaml_data_models.py [hunk 4]
```diff
@@ -541,11 +548,12 @@ class ConfigWorkflow(BaseModel):
 
     """
 
+    rootdir: None | Path = None
     name: str | None = None
-    cycles: list[ConfigCycle]
-    tasks: list[ConfigTask]
+    cycles: Annotated[list[ConfigCycle], AfterValidator(list_not_empty)]
+    tasks: Annotated[list[ConfigTask], AfterValidator(list_not_empty)]
```
Comment (DropD) [inline src/sirocco/parsing/_yaml_data_models.py:554]: If we can add the validators here, then the `CanonicalWorkflow` class only adds the rootdir. I agree that then, there is a more elegant solution that duplicates less code and is orthogonal to the changes to `workflow.Workflow` in this PR.

File: src/sirocco/core/workflow.py [hunk 5]
```diff
@@ -1,60 +1,77 @@
...
+
+        self.name: str = name
+        self.config_rootdir: Path = rootdir # TODO rename to rootdir since it does not make sense with this constructor
+
         self.tasks: Store = Store()
         self.data: Store = Store()
         self.cycles: Store = Store()
 
+        data_dict: dict[str, ConfigAvailableData | ConfigGeneratedData] = {data.name: data for data in chain(data.available, data.generated)}
+        task_dict: dict[str, ConfigTask] = {task.name: task for task in tasks}
```
Comment (DropD) [inline src/sirocco/core/workflow.py:45]: The properties could also be here (see the other comment about properties).

File: src/sirocco/core/workflow.py
Comment (DropD) [inline src/sirocco/core/workflow.py:1]: These changes strike me as orthogonal to the goal of canonicalization. They could be isolated and put into their own PR.

File: src/sirocco/core/workflow.py (hunk 5 above)
Comment (DropD) [reply src/sirocco/core/workflow.py:45]: or, for performance reasons, they could store the output of the properties.

Comment (agoscinski) [reply src/sirocco/core/workflow.py:45]: I think these are just temporaries that are needed for the creation of the Workflow. One can add them as properties but they are nowhere else in the code used

File: src/sirocco/parsing/_yaml_data_models.py (hunk 2 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:551]: It can be supplied by using the `from_config_file` constructor, it seems to me also reasonable that `ConfigWorkflow` is aware of its location. It is not perfect since you are right that the information is not supplied in the yaml file. The alternative, however, to make the `core.Workflow` aware of how the `ConfigWorkflow` should be used for parsing seems to me less ideal. So I think it is not the perfect but the alternative seems worse to me

File: tests/unit_tests/parsing/test_yaml_data_models.py (hunk 1 above)
Comment (agoscinski) [reply tests/unit_tests/parsing/test_yaml_data_models.py:8]: ok! will check this

File: src/sirocco/parsing/_yaml_data_models.py (hunk 3 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:598]: I just did not see where it is needed except for the conversion to `core.Workflow`.

File: src/sirocco/core/workflow.py
Comment (DropD) [reply src/sirocco/core/workflow.py:1]: I meant isolated from other changes, like undoing the shift to properties instead of duplicating information, moving the validators to the ConfigWorkflow and replacing CanonicalWorkflow with some other way of storing ConfigWorkflow + rootdir.

File: src/sirocco/core/workflow.py (hunk 5 above)
Comment (DropD) [reply src/sirocco/core/workflow.py:45]: Then properties are the way to go.

File: src/sirocco/parsing/_yaml_data_models.py (hunk 2 above)
Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:551]: In what way was `core.Workflow` aware of any of this in #89? Does it not provide exactly that alternative without the drawbacks?

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:554]: I would suggest replacing `CanonicalWorkflow` with a dataclass / pydantic model / namedtuple (whatever), which puts a type name to `ConfigWorkflow` + `rootdir`.  `workflow.Workflow` should then only deal with that type.

File: src/sirocco/parsing/_yaml_data_models.py (hunk 2 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:551]: You are right, this logic was in `_yaml_data_models` so `core.Workflow` was not aware of the parsing logic, I remembered it wrongly. Before the canonicalization the `rootdir` was updated in the  `_yaml_data_models.load_workflow_config` function. Then this was moved to the construction of the `CanonicalWorkflow`. In this PR we move it to the `ConfigWorkflow` into the new `from_config_file` constructor so it is handled within the class which is more logic than adding it an util function as it was before the canonicalization. I think this an implementation imitating the benefits of the `CanonicalWorkflow` but moving it to the `ConfigWorkflow`. One can use `rootdir` is an indicator if the workflow was created from a file or not.

File: src/sirocco/core/workflow.py (hunk 5 above)
Comment (agoscinski) [reply src/sirocco/core/workflow.py:45]: This I don't fully understand, why introduce a property for something that is computed once? You expose it to the API of the class. Do you want to capsulate the logic to test it? This I could understand, but then I would make them private properties.

File: src/sirocco/core/workflow.py
Comment (agoscinski) [reply src/sirocco/core/workflow.py:1]: I can split it up maybe in multiple commits?

File: src/sirocco/core/workflow.py (hunk 5 above)
Comment (agoscinski) [reply src/sirocco/core/workflow.py:45]: Hm.. no as property it also does not make sense as it is created from the `tasks: ConfigTask` that are only provided during initialization in `core.Workflow`, so accessing it afterwards does not make so much sense. If you want to test it I would propose to imitate the `build_internal_dicts` function in `core.Workflow`
```python
    def _build_data_dict(self, data: ConfigData) -> dict[str, ConfigAvailableData | ConfigGeneratedData]:
        return data_dict = {data.name: data for data in data.available} | {
            data.name: data for data in data.generated
        }
```
same for task dict

File: tests/unit_tests/parsing/test_yaml_data_models.py (hunk 1 above)
Comment (agoscinski) [reply tests/unit_tests/parsing/test_yaml_data_models.py]: ok done!

File: src/sirocco/core/workflow.py (hunk 5 above)
Comment (agoscinski) [reply src/sirocco/core/workflow.py]: I did it now like this, what do you think?

File: src/sirocco/core/workflow.py [hunk 6]
```diff
@@ -1,60 +1,80 @@
...
-        self.name: str = workflow_config.name
-        self.config_rootdir: Path = workflow_config.rootdir
+    def __init__(self,
+                 name: str,
+                 rootdir: Path,
+                 cycles: list[ConfigCycle],
+                 tasks: list[ConfigTask],
+                 data: ConfigData,
+                 parameters: dict[str, list], # TODO type should be more specified but not in this PR
+                 ) -> None:
```
Comment (DropD) [inline src/sirocco/core/workflow.py:35]: This signature change is valuable, especially together with the `from_config_workflow` classmethod.

File: src/sirocco/core/workflow.py [hunk 7]
```diff
@@ -1,60 +1,80 @@
...
+
+        self.name: str = name
+        if not isinstance(rootdir, Path):
+            msg = f"Need to provide path for rootdir but got {type(rootdir)}"
+            raise TypeError(msg)
```
Comment (DropD) [inline src/sirocco/core/workflow.py:40]: Moving this logic here is what we agreed not to do, as it is part of dealing with user input.

File: src/sirocco/core/workflow.py [hunk 8]
```diff
@@ -80,13 +100,38 @@ def iter_coordinates(param_refs: list, date: datetime | None = None) -> Iterator
         for task in self.tasks:
             task.link_wait_on_tasks(self.tasks)
 
+    @staticmethod
+    def _build_data_dict(data: ConfigData) -> dict[str, ConfigAvailableData | ConfigGeneratedData]:
+        return {data.name: data for data in chain(data.available, data.generated)}
+    @staticmethod
+    def _build_tasks_dict(tasks: list[ConfigTask]) -> dict[str, ConfigTask]:
+        return {task.name: task for task in tasks}
```
Comment (DropD) [inline src/sirocco/core/workflow.py:108]: If these were instance methods, there would be no need to pass `data` / `tasks` in there. Since they are prefixed with `_` you are declaring they should not be used outside an instance (where passing the information in would make sense). What is the design rationale for this contradiction?

File: src/sirocco/parsing/_yaml_data_models.py (hunk 2 above)
Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:551]: Please explain in which way moving `from_config_file` from the module to the class makes it "more logical"?

File: src/sirocco/parsing/_yaml_data_models.py [hunk 9]
```diff
@@ -570,67 +578,19 @@ def check_parameters(self) -> ConfigWorkflow:
...
-
-    rootdir = config_path.resolve().parent
-
-    return canonicalize_workflow(config_workflow=parsed_workflow, rootdir=rootdir)
-    # return parsed_workflow
+    @classmethod
+    def from_config_file(cls, config_filename: str):
+        # TODO maybe not best design, since we have multiple entry points, but
+        #      on the other hand we don't have rootdir ambiguity,I think this can be
+        #      better integrated with pydantic
```
Comment (DropD) [inline src/sirocco/parsing/_yaml_data_models.py:585]: In which way is it true that "we don't have rootdir ambiguity"? `.rootdir` is still type hinted as `Path | None` and a check in `workflow.Workflow.__init__` is still required, is there another definition of "ambiguity"?

File: src/sirocco/parsing/_yaml_data_models.py (hunk 2 above)
Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:551]: If we do this, this PR should add versions of each test that uses `ConfigWorkflow` with `rootdir=None` and check the appropriate outcome.

File: src/sirocco/parsing/_yaml_data_models.py [hunk 10]
```diff
@@ -541,11 +548,12 @@ class ConfigWorkflow(BaseModel):
 
     """
 
+    rootdir: None | Path = None
     name: str | None = None
```
Comment (DropD) [inline src/sirocco/parsing/_yaml_data_models.py:552]: Since now ambiguity in the type of `name` is reintroduced, all tests are required to have a version for `name=None`

File: src/sirocco/parsing/_yaml_data_models.py [hunk 11]
```diff
@@ -23,6 +23,15 @@
 
 from sirocco.parsing._utils import TimeUtils
 
+ITEM_T = typing.TypeVar("ITEM_T")
```
Comment (agoscinski) [inline src/sirocco/parsing/_yaml_data_models.py:26]: If the type is not specified why bother specifying it? Isn't list[ITEM_T] similar useful for type checking as list at the moment? Or will this type more explicitly specified in the future?

File: src/sirocco/core/workflow.py (hunk 7 above)
Comment (leclairm) [reply src/sirocco/core/workflow.py:40]: I second that

File: src/sirocco/parsing/_yaml_data_models.py (hunk 11 above)
Comment (leclairm) [reply src/sirocco/parsing/_yaml_data_models.py:26]: I presume that if it is used only once it doesn't make sense but when used for several input types or both input and output, it indicates those must be the same.

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (leclairm) [reply src/sirocco/parsing/_yaml_data_models.py:564]: > If we can add the validators here, then the `CanonicalWorkflow` class only adds the rootdir. I agree that then, there is a more elegant solution that duplicates less code and is orthogonal to the changes to `workflow.Workflow` in this PR.

 `CanonicalWorkflow` also makes `name` non-optional

File: src/sirocco/core/workflow.py (hunk 7 above)
Comment (agoscinski) [reply src/sirocco/core/workflow.py]: Okay removed, see latest commit

File: src/sirocco/parsing/_yaml_data_models.py (hunk 2 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py]: > Please explain in which way moving from_config_file from the module to the class makes it "more logical"? 

Since it is essentially a method that constructs an instance of the class.

> If we do this, this PR should add versions of each test that uses ConfigWorkflow with rootdir=None and check the appropriate outcome.

Added tests

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:564]: I used now a reasonable default argument for `rootdir` that I would have been favor for in a future PR, but since it simplifies the handling of the `rootdir` with both constructors, I introduced it here.

File: src/sirocco/parsing/_yaml_data_models.py (hunk 11 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:26]: Yes but we introduce it without doing this.

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:564]: If we move `rootdir` outside of the ConfigWorkflow I think it would become problematic with setting `rootdir` inside the yaml file

Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:564]: I realized that the user input in the file was not be respected so I had to to open up the parsing. I realized that the parsing of pydantic_yaml is actually not using pydantic v2 so I would remove it in a future PR. As a consequence the default argument for `rootdir` was not required anymore to get a technical simpler solution, so I removed it to reintroduce this in some later PR.

File: src/sirocco/parsing/_yaml_data_models.py (hunk 11 above)
Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:26]: Here is how to read `def list_not_empty(value: list[ITEM_T]) -> list[ITEM_T]:`:

- `list_not_empty` is a function `list[T] -> list[T]`, therefore
- it is allowed to take `[1, 2, 3]` and return `[4, 5, 6, 7, 8]
- it is not allowed to take `[1, 2, 3]` and return `["a", "b", "c"]`

The guarantee is that the mix of types in the input list is the same as in the output list

Here is how to read `def list_not_empty(value: list) -> list:`:

- `list_not_empty` is a function `list[Any] -> list[Any]`, therefore
- it is allowed to take `[1, 2, 3]` and return `["1", "2", "3"]`
- it is even allowed to take `[1, 2, 3]` and return `[("a", "b"), 4.5, None, [[[1], 2], 3]]`

The guarantee on the relationship of item types in the input and output list is absolutely none.

This is because `Any` is the union of all types.

Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:26]: This function is generic, but it is core to it's design that it does not change the input list. The closest we can do to letting type checkers know that is this. Now type checkers don't have to complain if we use `cycles` as a list of `Cycle`s after running this function on it. Otherwise it would have to assume that it can now contain anything.

Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:26]: > Isn't list[ITEM_T] similar useful for type checking as list at the moment?
In conclusion, no. It is a lot more useful, which will become super obvious when switching on type checking project wide in CI.

Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:26]: okay I get the difference to just use `list[Any]` thanks for the clarification

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (DropD) [reply src/sirocco/parsing/_yaml_data_models.py:572]: I had sort of assumed that there was a design decision to not include `rootdir` in the yaml config. If we can (and also make it and `name` mandatory) then `CanonicalWorkflow` can be removed without replacement.

File: src/sirocco/core/workflow.py [hunk 12]
```diff
@@ -88,5 +107,21 @@ def cycle_dates(cycle_config: ConfigCycle) -> Iterator[datetime]:
...
+        Loads a python representation of a workflow config file.
+
+        :param config_filename: the string to the config yaml file containing the workflow definition
+        """
+        return cls.from_config_workflow(ConfigWorkflow.from_config_file(config_filename))
+
+    @classmethod
+    def from_config_workflow(cls: type[Self], config_workflow: ConfigWorkflow) -> Workflow:
+        return cls(
+            name=config_workflow.name,
```
Comment (DropD) [inline src/sirocco/core/workflow.py:121]: This will fail mypy checks, because `config_workflow.name` can be `None`. That was the main job of `CanonicalWorkflow` to avoid.

File: src/sirocco/core/workflow.py [hunk 13]
```diff
@@ -88,5 +107,21 @@ def cycle_dates(cycle_config: ConfigCycle) -> Iterator[datetime]:
...
+            rootdir=config_workflow.rootdir,
```
Comment (DropD) [inline src/sirocco/core/workflow.py:122]: This will fail mypy checks, because `ConfigWorkflow.rootdir` can be None. That was the main job of CanonicalWorkflow to avoid.

File: src/sirocco/core/workflow.py (hunk 12 above)
Comment (DropD) [reply src/sirocco/core/workflow.py:121]: This PR needs to catch that now. In general, please run `mypy` manually and check that it isn't a problem anywhere. This is the bar for any PR that wants to replace `CanonicalWorkflow` (You can run mypy there and compare).

File: src/sirocco/core/workflow.py (hunk 13 above)
Comment (DropD) [reply src/sirocco/core/workflow.py:122]: See [the other comment about this](https://github.com/C2SM/Sirocco/pull/100/files#r1930575335).

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:572]: > I had sort of assumed that there was a design decision to not include rootdir in the yaml config

I am not sure. I think the name `rootdir` indicates that it should not be set by the user, but right now it acts more like a working directory which I would argue should be settable by the user. One could introduce a new variable that makes this difference apparent. On the other hand, the only the usage of `rootdir` is as a working directory, ergo one could replace it. What do you think @DropD @leclairm

File: src/sirocco/core/workflow.py (hunk 13 above)
Comment (agoscinski) [reply src/sirocco/core/workflow.py:122]: `rootdir` and `name` cannot be None anymore

File: src/sirocco/parsing/_yaml_data_models.py (hunk 10 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py]: `rootdir` and `name` cannot be None anymore

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (leclairm) [reply src/sirocco/parsing/_yaml_data_models.py:572]: `rootdir` is the directory where we find the full configuration, which consists not only of the yaml file but also all the other ones like ICON namelists, user scripts for shell tasks, etc... Subsequently, this is also the root path used to resolve the relative paths provided in the yaml file for these other config files. So it has to simply be the parent directory of the yaml file, nothing else, and must not be set by the user in the yaml file.

Regarding its name, `rootdir` seems logical as it is originally a property of `CanonicalWorkflow` being itself derived from `ConfigWorkflow`. Then it becomes `Workflow.config_rootdir` which still makes sense. The only name we could change here for consistency is `CanonicalWorkflow` -> `CanonicalConfigWorkflow` (if the canonical class is kept, I still need to read through the PR again to understand what it enables)

**Review by leclairm — CHANGES_REQUESTED** [review]

**Comment by leclairm** [comment]
I'm fine with the PR, once we've handled the `ConfigWorkflow.rootdir` issue (cannot be set by the user).

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:572]: > rootdir is the directory where we find the full configuration, which consists not only of the yaml file but also all the other ones like ICON namelists, user scripts for shell tasks, etc... Subsequently, this is also the root path used to resolve the relative paths provided in the yaml file for these other config files. So it has to simply be the parent directory of the yaml file, nothing else, and must not be set by the user in the yaml file.

But here again you argue that there is must be `rootdir`, then we use it as working directory. I argue there must be working directory, and if this working directory is not specified we use the `rootdir` as it. I don't see why we should not give the user the possibility to specify where the data for the workflow is so the user has more flexibility in arranging his/her folder structure.

File: src/sirocco/core/workflow.py (hunk 12 above)
Comment (agoscinski) [reply src/sirocco/core/workflow.py:121]: `name` cannot be anymore None, by separating the yaml parsing and validation one can set it in the `from_config_file` constructor. The only reason why we allowed `None` was because before we could not determine if `name` is specified in the given file.

File: src/sirocco/parsing/_yaml_data_models.py [hunk 14]
```diff
@@ -570,67 +596,21 @@ def check_parameters(self) -> ConfigWorkflow:
...
+
+        config_path = Path(config_filename)
+        content = config_path.read_text()
+        reader = YAML(typ="safe", pure=True)
+        object_ = reader.load(StringIO(content))
+        # If name was not specified, then we use filename without file extension
+        if "name" not in object_:
+            object_["name"] = config_path.stem
+        if "rootdir" not in object_:
+            object_["rootdir"] = config_path.resolve().parent
```
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:614]: I was first advocating for being able to manipulate the working directory, but when I thought this through I came to the conclusion that this not important (see issue https://github.com/C2SM/Sirocco/issues/111).

File: src/sirocco/parsing/_yaml_data_models.py [hunk 15]
```diff
@@ -533,17 +546,30 @@ class ConfigWorkflow(BaseModel):
...
 
     """
 
-    name: str | None = None
-    cycles: list[ConfigCycle]
-    tasks: list[ConfigTask]
+    rootdir: Path
+    name: str
+    cycles: Annotated[list[ConfigCycle], AfterValidator(list_not_empty)]
+    tasks: Annotated[list[ConfigTask], AfterValidator(list_not_empty)]
```
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:572]: I just replicated the existing check. In this case it does not make a real difference. If I am not wrong, it just specifies if it happens before the cycles are converted from list of dictionaries to a list of ConfigCycles or after. Doing it after or before should not change it logically. But yes we could change it to `BeforeValidator`.

File: src/sirocco/core/workflow.py [hunk 16]
```diff
@@ -1,60 +1,79 @@
...
 class Workflow:
     """Internal representation of a workflow"""
 
-    def __init__(self, workflow_config: CanonicalWorkflow) -> None:
-        self.name: str = workflow_config.name
-        self.config_rootdir: Path = workflow_config.rootdir
+    def __init__(
+        self,
+        name: str,
+        rootdir: Path,
```
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/core/workflow.py:31]: I think this will be changed once rebased. Waiting for PR #110 being merged

File: src/sirocco/parsing/_yaml_data_models.py [hunk 17]
```diff
@@ -570,67 +596,21 @@ def check_parameters(self) -> ConfigWorkflow:
...
-    # If name was not specified, then we use filename without file extension
-    if parsed_workflow.name is None:
-        parsed_workflow.name = config_path.stem
-
-    rootdir = config_path.resolve().parent
-
-    return canonicalize_workflow(config_workflow=parsed_workflow, rootdir=rootdir)
-    # return parsed_workflow
+    @classmethod
+    def from_config_file(cls, config_filename: str):
```
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:600]: depends, path could also mean that you need to specify a `pathlib.Path`

[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:600]: ah okay, yes in this case we actually specify a path not a filename, will do it

File: src/sirocco/core/workflow.py [hunk 18]
```diff
@@ -88,5 +107,21 @@ def cycle_dates(cycle_config: ConfigCycle) -> Iterator[datetime]:
                 yield date
 
     @classmethod
-    def from_yaml(cls: type[Self], config_path: str) -> Self:
-        return cls(load_workflow_config(config_path))
+    def from_yaml(cls: type[Self], config_filename: str) -> Self:
```
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/core/workflow.py:110]: yes I wanted to change this in another PR to keep it separate, but one could also do it in this PR and make another commit

File: src/sirocco/parsing/_yaml_data_models.py (hunk 4 above)
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py:572]: We decided in a meeting to keep `rootdir` as a name because `working_directory` implies too much that the calculation happens there

File: src/sirocco/core/workflow.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/core/workflow.py:110]:
```suggestion
    def from_ config_file(cls: type[Self], config_file_path: str) -> Self:
```

[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/core/workflow.py]: changed now in this PR

[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/core/workflow.py]: I used `config_path` since this name was used in the visualizer

File: src/sirocco/parsing/_yaml_data_models.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (agoscinski) [reply src/sirocco/parsing/_yaml_data_models.py]: changed, I used config_path since this name was used in the visualizer

File: tests/unit_tests/parsing/test_yaml_data_models.py [hunk 19]
```diff
@@ -63,6 +46,38 @@ def test_load_workflow_config(tmp_path):
...
+
+    with pytest.raises(ValidationError, match=r".*1 validation error for ConfigWorkflow\nrootdir.*"):
+        models.ConfigWorkflow(
+            name=minimal_config.name,
+            rootdir=None,
+            cycles=minimal_config.cycles,
+            tasks=minimal_config.tasks,
+            data=minimal_config.data,
+            parameters=minimal_config.parameters,
+        )
```
Comment (agoscinski) [inline tests/unit_tests/parsing/test_yaml_data_models.py:83]: TODO remove

**Review by DropD — APPROVED** [review]

File: tests/unit_tests/parsing/test_yaml_data_models.py (hunk 19 above)
Comment (agoscinski) [inline tests/unit_tests/parsing/test_yaml_data_models.py:83]:
```suggestion
```

**Review by leclairm — APPROVED** [review]
